### PR TITLE
VPLAY-12908 - Code improvement in TsbReader

### PR DESCRIPTION
**Pull request by hthakkar-synamedia**
Reason for change: - A follow up change to VPLAY-11126. Some edge cases are addressed here.
Test Procedure: - Refer to Ticket
Risks: low

### CODE REVIEW

File: fragmentcollector_mpd.cpp [hunk 1]
```diff
@@ -10048,15 +10066,9 @@ void StreamAbstractionAAMP_MPD::TsbReader()
...
-					// play cache is full , wait until cache is available to inject next, max wait of 1sec
-					int timeoutMs = MAX_WAIT_TIMEOUT_MS;
 					int trackIdx = (vEOS && !aEOS) ? eMEDIATYPE_AUDIO : eMEDIATYPE_VIDEO;
-					//AAMPLOG_INFO("Cache full state track(%d), no download until(%d) Time(%lld)",trackIdx,timeoutMs,aamp_GetCurrentTimeMS());
-					bool temp =  mMediaStreamContext[trackIdx]->WaitForCachedFragmentChunkInjected(timeoutMs);
-					if(temp == false)
-					{
-						//AAMPLOG_INFO("Waiting for FreeFragmentAvailable");  //CID:82355 - checked return
-					}
+					//// play cache is full , wait until cache is available to inject next, max wait of MAX_WAIT_TIMEOUT_MS
```
Comment (jfagunde) [inline fragmentcollector_mpd.cpp]: Too many '/'?

File: fragmentcollector_mpd.cpp [hunk 2]
```diff
@@ -10048,15 +10066,9 @@ void StreamAbstractionAAMP_MPD::TsbReader()
...
+					mMediaStreamContext[trackIdx]->WaitForCachedFragmentChunkInjected(MAX_WAIT_TIMEOUT_MS);
```
Comment (jfagunde) [inline fragmentcollector_mpd.cpp]: I would void this call, to make it clear that it is intentional to ignore the return value.

File: test/utests/tests/AampTSBSessionManager/FunctionalTests.cpp [hunk 3]
```diff
@@ -653,3 +664,66 @@ TEST_F(FunctionalTests, AdMetadataBoundaryTest)
...
+			.WillOnce(Return(static_cast<int>(TSB::LogLevel::TRACE)));
+		EXPECT_CALL(*g_mockAampTsbMetaDataManager, Initialize()).WillOnce(Return());
+		EXPECT_CALL(*g_mockAampTsbMetaDataManager,
+					RegisterMetaDataType(AampTsbMetaData::Type::AD_RESERVATION_METADATA_TYPE, false))
+			.WillOnce(Return(true));
+		EXPECT_CALL(*g_mockAampTsbMetaDataManager,
+					RegisterMetaDataType(AampTsbMetaData::Type::AD_PLACEMENT_METADATA_TYPE, true))
+			.WillOnce(Return(true));
+		partialMock.Init();
+		std::this_thread::sleep_for(std::chrono::milliseconds(25));
```
Comment (jfagunde) [inline test/utests/tests/AampTSBSessionManager/FunctionalTests.cpp]: Is this sleep needed? If it is, I'd add a comment explaining why.

File: test/utests/tests/AampTSBSessionManager/FunctionalTests.cpp [hunk 4]
```diff
@@ -653,3 +664,66 @@ TEST_F(FunctionalTests, AdMetadataBoundaryTest)
...
+		// Key assertion: the real Flush() must call NotifyVideoTsbWaiters() exactly once.
+		EXPECT_CALL(partialMock, NotifyVideoTsbWaiters()).Times(1);
+		EXPECT_CALL(*g_mockTSBStore, Flush()).Times(1);
+		partialMock.Flush();
+		// ~PartialMockTSBSessionManager() runs here, verifying the EXPECT_CALL above.
+		// ~AampTSBSessionManager() then calls Flush(), but mInitialized_ is false so
+		// NotifyVideoTsbWaiters() is not invoked.
+	}
+
+	// Re-create a plain instance so TearDown can perform its own Flush cleanly.
```
Comment (jfagunde) [inline test/utests/tests/AampTSBSessionManager/FunctionalTests.cpp]: Is this needed? Why do we need TearDown to perform its own Flush cleanly?

File: AampTSBSessionManager.cpp
Comment (jfagunde) [inline AampTSBSessionManager.cpp:380]: How about adding a TRACE when the cond var is notified?

Comment (jfagunde) [inline AampTSBSessionManager.cpp:388]: How about printing a TRACE just before waiting and another one after wake up?
Previously, when running out of segments in the TSB, we were printing a TRACE every 50ms. This would only print two: one when starting to wait and another one when waking up. It would help identify problems in which the cond var is not notified for whatever reason.

File: fragmentcollector_mpd.cpp (hunk 1 above)
Comment (hthakkar-synamedia) [reply fragmentcollector_mpd.cpp]: Removed extra

File: fragmentcollector_mpd.cpp (hunk 2 above)
Comment (hthakkar-synamedia) [reply fragmentcollector_mpd.cpp]: Done

File: AampTSBSessionManager.cpp
Comment (hthakkar-synamedia) [reply AampTSBSessionManager.cpp:380]: Done

Comment (hthakkar-synamedia) [reply AampTSBSessionManager.cpp:388]: Done

File: test/utests/tests/AampTSBSessionManager/FunctionalTests.cpp (hunk 3 above)
Comment (hthakkar-synamedia) [reply test/utests/tests/AampTSBSessionManager/FunctionalTests.cpp]: Init() starts the write thread (mWriteThread) asynchronously. The sleep gives that thread time to reach its blocking wait() call inside ProcessWriteQueue(). Without it, Flush() could run before the thread is actually waiting on the condition variable, causing a race where notify_one() fires before the thread is ready to receive it — potentially causing mWriteThread.join() to hang or the thread to miss the stop signal and not exit cleanly.

File: test/utests/tests/AampTSBSessionManager/FunctionalTests.cpp (hunk 4 above)
Comment (hthakkar-synamedia) [reply test/utests/tests/AampTSBSessionManager/FunctionalTests.cpp]: Simplified code to remove this

**Review by jfagunde — APPROVED** [review]

File: test/utests/tests/AampTSBSessionManager/FunctionalTests.cpp [hunk 5]
```diff
@@ -653,3 +658,64 @@ TEST_F(FunctionalTests, AdMetadataBoundaryTest)
 	EXPECT_TRUE(mAampTSBSessionManager->StartAdPlacement(
 		TEST_AD_ID, TEST_REL_POSITION, zeroPos, 0.0, TEST_OFFSET));
 }
+
+/**
+ * @brief Verify that Flush() calls NotifyVideoTsbWaiters().
+ *
+ * Uses a stack-allocated plain AampTSBSessionManager so the real Flush() and
```
Comment (jfagunde) [inline test/utests/tests/AampTSBSessionManager/FunctionalTests.cpp]: Why? Can we not use mAampTSBSessionManager instead?

**Review by jfagunde — APPROVED** [review]
